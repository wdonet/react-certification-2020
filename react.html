<html>
  <head>
    <title>MarioTube</title>
  </head>
  <body>
    <!--
        <div class="header">
            <div class="header-logo"><img src="./m-vector_icon.jpg" /></div>
            <div class="header-search">
                <label>Search videos <input type="text" name="search" placeholder="ðŸŽ¥  Search for videos"/></label>
            </div>
            <div class="header-login">
                <a>Sign in</a> <img src="./wize-logo.png" />
            </div>
        </div>
        <div class="sidebar">
          <ul>
            <li>Home</li>
            <li>Favorites</li>
          </ul>
          <div>Dark Mode</div>
        </div>
        <div class="content">
          <div class="categories">Categories</div>
          <div class="videos"><div class="video"></div></div>
        </div>
        <div class="footer"><div class="info"></div></div>

/**
* Sample JavaScript code for youtube.search.list
* See instructions for running APIs Explorer code samples locally:
* https://developers.google.com/explorer-help/guides/code_samples#javascript
*/-->
    <script src="https://apis.google.com/js/api.js"></script>
    <script>
      function authenticate() {
        console.log('authenticate...');
        return gapi.auth2
          .getAuthInstance()
          .signIn({ scope: 'https://www.googleapis.com/auth/youtube.force-ssl' })
          .then(
            function () {
              console.log('Sign-in successful');
            },
            function (err) {
              console.error('Error signing in', err);
            }
          );
      }
      function loadClient() {
        console.log('loadClient...');
        console.log('--setApiKey...');
        gapi.client.setApiKey('AIzaSyDpytkLvHqG_bnDtwfSklDRk-eU8O-fHL8');
        return gapi.client
          .load('https://www.googleapis.com/discovery/v1/apis/youtube/v3/rest')
          .then(
            function () {
              console.log('GAPI client loaded for API');
            },
            function (err) {
              console.error('Error loading GAPI client for API', err);
            }
          );
      }
      // Make sure the client is loaded and sign-in is complete before calling this method.
      function execute() {
        console.log('execute...');
        console.log('--search.list...');
        return gapi.client.youtube.search
          .list({
            part: ['snippet'],
            q: 'lacuna coil',
          })
          .then(
            function (response) {
              // Handle the results here (response.result has the parsed body).
              console.log('--Response', response);
            },
            function (err) {
              console.error('--Execute error', err);
            }
          );
      }
      gapi.load('client:auth2', function () {
        console.log('load...');
        console.log('--client:auth2...');
        console.log('--init...');
        gapi.auth2.init({
          client_id:
            '1088412977008-bt87ku9si9u48echferpm6i23rcv5jei.apps.googleusercontent.com',
        });
      });
    </script>
    <button onclick="authenticate().then(loadClient)">authorize and load</button>
    <button onclick="execute()">execute</button>

    <!--

    html, body, #root {
        width: 100%;
        height: 100%;
      }
      
      .SplitPane {
        width: 100%;
        height: 100%;
      }
      
      .SplitPane-left {
        float: left;
        width: 10%;
        height: 100%;
      }
      
      .SplitPane-right {
        float: left;
        width: 89%;
        height: 100%;
      }
      
      .Contacts {
        width: 100%;
        height: 100%; 
        background: lightblue;
      }
      
      .Chat {
        width: 100%;
        height: 100%; 
        background: pink;
      }


      function Contacts() {
        return <div className="Contacts" />;
      }
      
      function Chat() {
        return <div className="Chat" />;
      }
      
      function SplitPane(props) {
        return (
          <div className="SplitPane">
            <div className="SplitPane-left">
              {props.left}
            </div>
            <div className="SplitPane-right">
              {props.right}
            </div>
          </div>
        );
      }
      
      function App() {
        return (
          <SplitPane
            left={
              <Contacts />
            }
            right={
              <Chat />
            } />
        );
      }
      
      ReactDOM.render(
        <App />,
        document.getElementById('root')
      );
      

      YouTube App


      Menu Bar
          Icon    
          Search bar
          Login icon
              Login
              Account info
              Logout
      
      Side Bar
          Home
          Favorites
          Dark mode
      
      Main Body
          Videos
              Thumbnail
              Title
              Description
              Author
      
      Video View
          Video
              Title               Views
              Author              Like/Dislike
              Description
      
      Footer
      
      KNEWZ-1745
      
      React.createElement(
          "div",
          {},
          React.createElement("h1", {}, props.title),
          React.createElement("h1", {}, props.subtitle),
      )


      class LoginControl extends React.Component {
        constructor(props) {
            super(props);
            this.handleLoginClick = this.handleLoginClick.bind(this);
            this.handleLogoutClick = this.handleLogoutClick.bind(this);
            this.state = {isLoggedIn: false};
        }
      
        handleLoginClick() {
            this.setState({isLoggedIn: true});
        }
      
        handleLogoutClick() {
            this.setState({isLoggedIn: false});
        }
      
        render() {
            const isLoggedIn = this.state.isLoggedIn;
            let button;
      
            if (isLoggedIn) {
                button = <LogoutButton onClick={this.handleLogoutClick} />;
            } else {
                button = <LoginButton onClick={this.handleLoginClick} />;
            }
      
            return (
                <div>
                    <Greeting isLoggedIn={isLoggedIn} />
                    {button}
                </div>
            );
        }
    }
      
    function UserGreeting(props) {
        return <h1>Welcome back!</h1>;
    }
      
    function GuestGreeting(props) {
        return <h1>Please sign up.</h1>;
    }
      
    function Greeting(props) {
        const isLoggedIn = props.isLoggedIn;
        if (isLoggedIn) {
            return <UserGreeting />;
        }
        return <GuestGreeting />;
    }
      
    function LoginButton(props) {
        return (
            <button onClick={props.onClick}>
                Login
            </button>
        );
    }
      
    function LogoutButton(props) {
        return (
            <button onClick={props.onClick}>
                Logout
            </button>
        );
    }
      
    ReactDOM.render(
        <LoginControl />,
        document.getElementById('root')
    );
      
-->
  </body>
</html>
