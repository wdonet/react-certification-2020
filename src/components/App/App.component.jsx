import React from 'react';
import { BrowserRouter, Switch, Route } from 'react-router-dom';

import HomePage from '../../views/Home';
import Header from '../Navbar';

function App() {

  return (
    <BrowserRouter>
      <Header />
      <Switch>
        <Route exact path="/">
          <HomePage />
        </Route>
      </Switch>
    </BrowserRouter>
  );
}

export default App;
