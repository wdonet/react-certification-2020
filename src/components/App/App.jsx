import React from 'react';
import NavBar from '../NavBar';
import HomePage from '../../pages/Home';

function App() {
  return (
    <div>
      <NavBar />
      <HomePage />
    </div>
  );
}

export default App;
