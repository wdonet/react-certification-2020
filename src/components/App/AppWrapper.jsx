import React, { useReducer } from 'react';
import { appReducer } from '../../reducers/appReducer';
import { lightTheme, darkTheme } from '../../providers/themes';
import youtubeSearchService from '../../services/youtubeSearchService';
import AppContext from '../../providers/AppContext';
import { 
  SET_IS_FIRST_LOAD, 
  SET_SESSION, 
  SET_VIDEOS_LIST, 
  SWITCH_THEME 
} from '../../reducers/actionTypes';

const USER_SESSION_KEY = "user_session";
const initialState = {
  videosList: [],
  currentVideoId: undefined,
  isLightTheme: true,
  isFirstLoad: true,
  userSession: JSON.parse(window.sessionStorage.getItem(USER_SESSION_KEY)),
};

function AppWrapper(props) {
  const [{ 
    videosList, 
    isLightTheme, 
    isFirstLoad,
    userSession}, 
    dispatch] = useReducer(appReducer, initialState);

  const getThemeConfig = () => ({
    theme: isLightTheme ? lightTheme : darkTheme,
    switchTheme: () => dispatch({type: SWITCH_THEME}),
  });

  const search = async (query) => {
    const serviceResponse = await youtubeSearchService.search(query);
    serviceResponse.error 
    ? dispatch({ type: SET_VIDEOS_LIST, payload: [] })
    : dispatch({ type: SET_VIDEOS_LIST, payload: serviceResponse });
  }

  const firstSearch = async () => {
    let counter = 0;
    /* global gapi */
    /* eslint no-undef: "error" */
    while((!gapi || !gapi.client) && counter < 10){
      counter+=1;
      await new Promise(resolve => setTimeout(resolve, 100));
    }
    ((!gapi || !gapi.client) || counter >= 10) 
    ? dispatch({ type: SET_VIDEOS_LIST, payload: [] })
    : search()
  }

  if(isFirstLoad){
    dispatch({ type: SET_IS_FIRST_LOAD, payload: false });
    firstSearch();
  }

  return (
    <AppContext.Provider value={
      { ...getThemeConfig(), 
        search,
        videosList,
        userSession,
        setUserSession: (data) => dispatch({type: SET_SESSION, payload: data}),
      }
    }>
      { props.children }
    </AppContext.Provider>
  );
}

export default AppWrapper;
