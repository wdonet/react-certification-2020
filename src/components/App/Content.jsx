import React from 'react';
import { Redirect, Route, Switch } from 'react-router';
import HomePlayer from '../HomeVideos/HomePlayer';
import HomeVideos from '../HomeVideos/HomeVideos';

const Content = () => {
    return (<div>
        <Switch>
            <Route exact path="/">
                <Redirect to="/home"/>
            </Route>
            <Route path="/home">
                <HomeVideos />
            </Route>
            <Route path="/player">
                <HomePlayer />
            </Route>
        </Switch>
    </div>);
}

export default Content;