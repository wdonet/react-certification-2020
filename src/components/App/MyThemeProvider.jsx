import React from 'react';
import { ThemeProvider } from 'styled-components';
import { useYouTube } from '../YouTube/YouTubeProvider';

const theme = {
  dark: {
    primary: '#4396f3',
    secondary: '#4a4646',
    color: '#f7f7f7',
  },
};

function MyThemeProvider({ children }) {
  const { state } = useYouTube();
  const { theme: stateTheme } = state;
  return <ThemeProvider theme={theme[stateTheme] || {}}>{children}</ThemeProvider>;
}

export default MyThemeProvider;
