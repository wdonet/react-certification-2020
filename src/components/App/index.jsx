import React, { useEffect, useState }from 'react';
import HeaderBar from '../HeaderBar';
import Content from '../Content';
import { items } from '../../assets/mockdata/mockdata.json';
import VideoCard from '../VideoCard';
import filterByYear from '../../utils/filter';

function App() {
  const [recent, setRecent] = useState([]);
  const [inDetail, setInDetail] = useState(false);
  const [detailVideoId, setDetailVideoId] = useState("");
  const [detailTitle, setDetailTitle] = useState("");
  const [detailDescription, setDetailDescription] = useState("");
  
  
  useEffect(() => { 
      console.log('useEffect');
      fetch(`https://www.googleapis.com/youtube/v3/search?key=${process.env.REACT_APP_YOUTUBE_API_KEY2}&q=wizeline&part=snippet&maxResults=50&order=date&type=video`)
      .then((res) => res.json())
      .then((res) => {
        if (res.items) {
          setRecent(res.items);
        } else {
          setRecent(items);
        }
        
      }
    ).catch((emAll) => {});
  }, []);
  
  const UpdateQuery = (text) => {
    console.log(`New query: ${text}`);
    fetch(`https://www.googleapis.com/youtube/v3/search?key=${process.env.REACT_APP_YOUTUBE_API_KEY2}&q=${text}&part=snippet&maxResults=50&order=date&type=video`)
      .then((res) => res.json())
      .then((res) => {
        if (res.items) {
          setRecent(res.items);
        } else {
          setRecent(items);
        }
        setInDetail(false);
      });
  };

  const changeDetail = (indetail=false, videoId, title, description) => {
    // setInDetail(inDetail ? false : true);
    console.log(`title: ${title}`);
    console.log(`description: ${description}`);
    setInDetail(indetail);
    setDetailVideoId(videoId);
    setDetailTitle(title);
    setDetailDescription(description);
  };


  return (
    <React.StrictMode>
      <HeaderBar updateQuery={ UpdateQuery } />
      <Content inDetail={inDetail} detailVideoId={detailVideoId} gotodetail={changeDetail} detailTitle={detailTitle} detailDescription={ detailDescription}>{recent ?
        recent.map((video) => {
          return (
            <VideoCard
              key={video.etag}
              videoId ={video.id.videoId}
              title={video.snippet.title}
              description={video.snippet.description}
              image={video.snippet.thumbnails.high.url}
              gotodetail={ changeDetail }
            />
          )
        })
      :null}
      </Content>
    </React.StrictMode>
  );
}

export default App;
