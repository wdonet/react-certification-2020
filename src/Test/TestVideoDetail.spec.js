import React from 'react';
import { render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { AppContext } from '../components/App/App.component';
import VideoDetailViewPage from '../pages/VideoDetailView/VideoDetailView.page';

describe('VideoDetailViewPage Components', () => {
  const VideoArray = [
    {
      id: {
        videoId: '123',
      },
      getVideoId: 'cHm3wVzcHjnE',
      snippet: {
        title: 'rick roll',
        thumbnails: {
          medium: {
            url: 'https://i.ytimg.com/vi/Po3VwR_NNGk/mqdefault.jpg',
          },
        },
      },
    },
  ];

  it('Take VideoDetailViewPage Snapshot ', () => {
    const { container } = render(
      <BrowserRouter>
        <AppContext.Provider value={false}>
          <VideoDetailViewPage detailVideo={VideoArray} />
        </AppContext.Provider>
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
  });

  it('Expect IFrame', () => {
    render(
      <BrowserRouter>
        <AppContext.Provider value={false}>
          <VideoDetailViewPage detailVideo={VideoArray} />
        </AppContext.Provider>
      </BrowserRouter>
    );
    const iframe = document.querySelector('iframe');
    expect(iframe).toBeTruthy();
  });

  it('Expect Card Button', () => {
    render(
      <BrowserRouter>
        <AppContext.Provider value={false}>
          <VideoDetailViewPage detailVideo={VideoArray} />
        </AppContext.Provider>
      </BrowserRouter>
    );
    expect(screen.getByRole('button')).toBeTruthy();
  });

  it('Expect Card Img', () => {
    render(
      <BrowserRouter>
        <AppContext.Provider value={false}>
          <VideoDetailViewPage detailVideo={VideoArray} />
        </AppContext.Provider>
      </BrowserRouter>
    );
    expect(screen.getByRole('img')).toBeTruthy();
  });

  it('Expect Card Title', () => {
    render(
      <BrowserRouter>
        <AppContext.Provider value={false}>
          <VideoDetailViewPage detailVideo={VideoArray} />
        </AppContext.Provider>
      </BrowserRouter>
    );
    const h3 = document.querySelector('h3');
    expect(h3).toBeTruthy();
  });
});
