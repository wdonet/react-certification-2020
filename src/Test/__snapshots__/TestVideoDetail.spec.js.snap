// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoDetailViewPage Components Take VideoDetailViewPage Snapshot  1`] = `
<div>
  <section
    style="position: absolute; top: 80px; width: 100%;"
  >
    <div
      class="sc-gKsewC iJqbzq"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <div
            class="sc-iBPRYJ kIKhxh"
          >
            <img
              alt=""
              class="sc-fubCfw FECHG"
              src="https://i.ytimg.com/vi/Po3VwR_NNGk/mqdefault.jpg"
            />
            <div
              class="sc-pFZIQ eDzivD"
            >
              rick roll
            </div>
          </div>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="sc-jrAGrp eDtBhv"
    >
      <div
        class="sc-jrAGrp eDtBhv"
      >
        <iframe
          allow="accelerometer; autoplay;"
          src="https://www.youtube.com/embed/undefined?controls=1&autoplay=1"
          style="width: 70%; display: flex; height: 450px; margin-top: 10px;"
          title="YouTubeAPI"
        />
      </div>
      <h3
        class="sc-kEjbxe cfhUBp"
      />
    </div>
  </section>
</div>
`;
